### PR TITLE
refactor easy_openyurt

### PR DESCRIPTION
**Pull request by lrq619**
## Summary

Refactor [easy_openyurt](https://github.com/flyinghorse0510/easy_openyurt) into vHive

## Implementation Notes :hammer_and_pick:

* Implement a new package `node` to execute all remote commands through ssh
* `clean` 's functionality is currently not applicable, need to reload all instances after each deployment
* demo is now conducted mannually and needs to automate

## External Dependencies :four_leaf_clover:

* N/A

## Breaking API Changes :warning:

* N/A

*Simply specify none (N/A) if not applicable.*

### CODE REVIEW

**Comment by lrq619** [comment]
@jchua99 Please take a look at this branch, where I refactor easy_openyurt into vHive using go scripts, there're still some tasks for you to do:
1. Take a look at the `clean` function, which is implemented in `main.go: func cleanNodes()`, it will clean up setups in all the nodes so that the deployment can be simply re-run. However, its functionality is not working properly, currently the re-deployment will get stuck in `yurt.go:209` after clean , please try to make it work as expected.
2. Demo is now conducted manually, please add a `demo.go` to automate this procedure as described in README.md

File: scripts/openyurt_deployer/configs/knative.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	LocalRegistryRepoVolumeSize          string
+	LocalRegistryVolumeConfigUrl         string
+	LocalRegistryDockerRegistryConfigUrl string
+	LocalRegistryHostUpdateConfigUrl     string
+	MagicDNSConfigUrl                    string
+	VHiveMode                            bool
+}
+
+var Knative = KnativeConfigStruct{
+	KnativeVersion:           "1.9.2",
```
Comment (ustiugov) [inline scripts/openyurt_deployer/configs/knative.go:21]: these configs should be taken from the vHive json config. there must be a consistent set of configs

File: scripts/openyurt_deployer/configs/kube.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+	ApiserverAdvertiseAddress string
+	PodNetworkCidr            string
+	PodNetworkAddonConfigURL  string
+	ApiserverPort             string
+	ApiserverToken            string
+	ApiserverTokenHash        string
+}
+
+var Kube = KubeConfigStruct{
+	K8sVersion:                "1.25.9",
```
Comment (ustiugov) [inline scripts/openyurt_deployer/configs/kube.go:15]: same here

File: scripts/openyurt_deployer/configs/system.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+// Current system environment
+var System = SystemEnvironmentStruct{
+	GoInstalled:                         false,
+	ContainerdInstalled:                 false,
+	RuncInstalled:                       false,
+	CniPluginsInstalled:                 false,
+	SystemdStartUp:                      true,
+	GoVersion:                           "1.18.10",
```
Comment (ustiugov) [inline scripts/openyurt_deployer/configs/system.go:42]: same for this config. also make sure the golang version is the same as in the CI

File: scripts/openyurt_deployer/configs/system.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	GoDownloadUrlTemplate:               "https://go.dev/dl/go%s.linux-%s.tar.gz",
+	ContainerdVersion:                   "1.6.18",
+	ContainerdDownloadUrlTemplate:       "https://github.com/containerd/containerd/releases/download/v%s/containerd-%s-linux-%s.tar.gz",
+	ContainerdSystemdProfileDownloadUrl: "https://raw.githubusercontent.com/containerd/containerd/main/containerd.service",
+	RuncVersion:                         "1.1.4",
+	RuncDownloadUrlTemplate:             "https://github.com/opencontainers/runc/releases/download/v%s/runc.%s",
+	CniPluginsVersion:                   "1.2.0",
+	CniPluginsDownloadUrlTemplate:       "https://github.com/containernetworking/plugins/releases/download/v%s/cni-plugins-linux-%s-v%s.tgz",
+	KubectlVersion:                      "1.25.9-00",
```
Comment (ustiugov) [inline scripts/openyurt_deployer/configs/system.go:51]: avoid duplicate entries for the k8s version

File: scripts/openyurt_deployer/go.mod [hunk 5]
```diff
@@ -0,0 +1,13 @@
+module github.com/vhive-serverless/vhive/scripts/openyurt_deployer
+
+go 1.20
```
Comment (ustiugov) [inline scripts/openyurt_deployer/go.mod:3]: the CI uses 1.19 IIRC

File: scripts/openyurt_deployer/configs/yurt.go [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+	WorkerAsEdge                    bool
+	Dependencies                    string
+	YurtVersion                     string
+}
+
+var Yurt = YurtEnvironment{
+	HelmInstalled:                   false,
+	HelmPublicSigningKeyDownloadUrl: "https://baltocdn.com/helm/signing.asc",
+	KustomizeInstalled:              false,
+	KustomizeScriptDownloadUrl:      "https://raw.githubusercontent.com/kubernetes-sigs/kustomize/master/hack/install_kustomize.sh",
```
Comment (ustiugov) [inline scripts/openyurt_deployer/configs/yurt.go:19]: what is this?

File: scripts/openyurt_deployer/configs/yurt.go [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+	MasterAsCloud:                   true,
+	WorkerNodeName:                  "",
+	WorkerAsEdge:                    true,
+	Dependencies:                    "",
+	YurtVersion:                     "1.2.1",
```
Comment (ustiugov) [inline scripts/openyurt_deployer/configs/yurt.go:24]: should be in a JSON config

File: scripts/openyurt_deployer/go.work [hunk 8]
```diff
@@ -0,0 +1,9 @@
+go 1.19
```
Comment (ustiugov) [inline scripts/openyurt_deployer/go.work:1]: please make sure all golang versions are the same

File: scripts/openyurt_deployer/logs/logs.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	_colorReset  = "\033[0m"
+	_colorRed    = "\033[31m"
+	_colorGreen  = "\033[32m"
+	_colorYellow = "\033[33m"
+	_colorBlue   = "\033[34m"
+	// _colorPurple = "\033[35m"
```
Comment (ustiugov) [inline scripts/openyurt_deployer/logs/logs.go:17]: uncomment or clean up

File: scripts/openyurt_deployer/main.go [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+	masterNode.InstallKnativeServing()
+	masterNode.InstallKnativeEventing()
+	logs.SuccessPrintf("Knative has been installed!\n")
+
+	logs.SuccessPrintf(">>>>>>>>>>>>>>>>OpenYurt Cluster Deployment Finished!<<<<<<<<<<<<<<<\n")
+
+
+}
+
+func demo(deployerConfFile string){
```
Comment (ustiugov) [inline scripts/openyurt_deployer/main.go:221]: demo-related stuff should be separate from the general tool

File: scripts/openyurt_deployer/node/demo.go [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+	"github.com/vhive-serverless/vhive/scripts/openyurt_deployer/template"
+)
+
+func (masterNode *Node) BuildDemo(workerNodes []Node) {
+
+	masterNode.GetUserHomeDir()
+	masterNode.GetNodeHostName()
+
+	var err error
+	// cloud.yaml
```
Comment (ustiugov) [inline scripts/openyurt_deployer/node/demo.go:17]: elaborate

File: scripts/openyurt_deployer/node/demo.go [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+	fmt.Printf("+Cloud Nodepool %s:\n", cloudPoolName)
+	fmt.Printf("+\t\t|_deployment: %s \n", deployments[0])
+	fmt.Printf("+\tnode: %s <- Master\n", masterNode.Configs.System.NodeHostName)
+	for _, worker := range workerNodes {
+		if worker.NodeRole == "cloud"{
+			fmt.Printf("+\tnode: %s\n", worker.Configs.System.NodeHostName)
+		}
+	}
+
+	fmt.Printf("+Edge  Nodepool %s:\n", edgePoolName)
```
Comment (ustiugov) [inline scripts/openyurt_deployer/node/demo.go:87]: don't use `fmt`, use the logger library from Haoyuan (or the generic logger we use in the core vHive)

File: scripts/openyurt_deployer/node/knative.go [hunk 13]
```diff
@@ -0,0 +1,158 @@
+package node
+
+import (
+	"github.com/vhive-serverless/vhive/scripts/openyurt_deployer/logs"
+)
+
+// func (node *Node) ParseSubcommandKnative(args []string) {
```
Comment (ustiugov) [inline scripts/openyurt_deployer/node/knative.go:7]: clean up the commented code

File: scripts/openyurt_deployer/node/knative.go [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+// 	}
+
+// 	node.InstallKnativeServing()
+// 	node.InstallKnativeEventing()
+
+// 	logs.SuccessPrintf("Init Knative Successfully! (vHive mode: %s)\n", vHiveMode)
+// }
+
+// Install Knative Serving
+func (node *Node) InstallKnativeServing() {
```
Comment (ustiugov) [inline scripts/openyurt_deployer/node/knative.go:54]: these scripts are duplicates of the ones in vHive. also make sure that you use the same Knative configs (and the same versions!)

File: scripts/openyurt_deployer/node/system.go [hunk 15]
```diff
@@ -0,0 +1,359 @@
...
+
+	// Disable swap
+	logs.InfoPrintf("Disabling swap")
+	_, err = node.ExecShellCmd("sudo swapoff -a && sudo cp /etc/fstab /etc/fstab.old") // Turn off Swap && Backup fstab file
+	logs.CheckErrorWithTagAndMsg(err, "Failed to disable swap!\n")
+
+	logs.InfoPrintf("Modifying fstab")
+	// Modify fstab to disable swap permanently
+	_, err = node.ExecShellCmd("sudo sed -i 's/#\\s*\\(.*swap.*\\)/\\1/g' /etc/fstab && sudo sed -i 's/.*swap.*/# &/g' /etc/fstab")
+	logs.CheckErrorWithTagAndMsg(err, "Failed to dodify fstab!\n")
```
Comment (ustiugov) [inline scripts/openyurt_deployer/node/system.go:244]: typo

File: scripts/openyurt_deployer/template/template.go [hunk 16]
```diff
@@ -0,0 +1,3 @@
+package template
+
+var vHiveConfigsURL = "https://raw.githubusercontent.com/anshalshukla/vHive/release-1.9/configs"
```
Comment (ustiugov) [inline scripts/openyurt_deployer/template/template.go:3]: I don't know what it is, but we should have everything in the repo, please fix

File: scripts/openyurt_deployer/utils.go [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+type ShellError struct {
+	msg      string
+	exitCode int
+}
+
+func (err *ShellError) Error() string {
+	return fmt.Sprintf("[exit %d] -> %s", err.exitCode, err.msg)
+}
+
+func ExecShellCmd(cmd string, pars ...any) (string, error) {
```
Comment (ustiugov) [inline scripts/openyurt_deployer/utils.go:33]: I am quite sure these utils are the same as in the vHive scripts.

**Review by ustiugov — CHANGES_REQUESTED** [review]
- [ ] fix inconsistent golang versions
- [ ] re-factor config parameters with appropriate JSON configs
- [ ] lint all the golang code (run `gofmt`)
- [ ] demo code should be separate from the code of the core/general toolchain. it is just a scenario script wich runs the general tool/scripts
- [ ] don't use `fmt.Print`, use the logger module
- [ ] add the vHive copyrights
- [ ] avoid script code duplication (e.g., for Knative setup, bash utils)
- [ ] replace YAML inlining into Golang files. Instead, have the YAML files and import them from the code

**Comment by leokondrashov** [comment]
@lrq619 What is the status of this PR?

**Comment by lrq619** [comment]
> @lrq619 What is the status of this PR?

Waiting for a fyp to modify it, he will submit a PR tmr, I will handle it then

**Comment by lrq619** [comment]
@jchua99 Hello, please fix these testing failures on this branch

**Comment by leokondrashov** [comment]
@lrq619 What is the status of this PR?

**Comment by lrq619** [comment]
> @lrq619 What is the status of this PR?

Should be closed now, Jason would push new PRs to replace this one

**Pull request closed without merging**